### PR TITLE
multi-DB Part 2: C++ interface API changes / swsscommon unit test / LOGLEVEL_DB apply new API

### PR DESCRIPTION
**Pull request by dzhangalibaba**
* add a new RedisConfig class which has static member and func, to parse/store/update/get database info from database_config.json file.

* add new DBConnector() constructor without port and unixsocket parameters. using RedisConfig class get func to get those info dynamically, all other DBConnector logic are same.

* the previous static sha member in tables are not good for multi-db cases, since on each db instance, there should be a SHA value, we cannot reuse only one SHA value, hence remove  the static keyword and keep SHA value as a member for each DBConnector object

* add updateDBConfigFile() to support vs test/ swsscommon test cases where this is no database_config.json at init time, we can point to load the config file.

* adding swsscommon unit test to check the  updateDBConfigFile() func, also check RedisConfig get APIs could return correct db info.

* relpace LOGLEVEL_DB with new DBConnector construct, LOGLEVEL_DB only has three places to changes, they are also in swsscommon repo, hence change here.  Since the database_config.json has only one redis inst for now. LOGLEVEL_DB behaves the same as before , though it is using new DBConnector constructor. All the other DBConnector replacement will be done in other PR.

* swsscommon test result ：
dzhang@343ca72c4bf5:/sonic/src/sonic-swss-common/tests$ ./tests --gtest_filter=DBConnector.multi_db_test
Running main() from gtest_main.cc
Note: Google Test filter = DBConnector.multi_db_test
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DBConnector
[ RUN      ] DBConnector.multi_db_test
Default : load init db config file, isInit = 0
Update : load local db config file, isInit = 1
JSON file # dbid: 0 ,socket: /var/run/redis/redis.sock .port: 6379
API get   # dbid: 0 ,socket: /var/run/redis/redis.sock .port: 6379
JSON file # dbid: 1 ,socket: /var/run/redis/redis1.sock .port: 6380
API get   # dbid: 1 ,socket: /var/run/redis/redis1.sock .port: 6380
JSON file # dbid: 4 ,socket: /var/run/redis/redis2.sock .port: 6381
API get   # dbid: 4 ,socket: /var/run/redis/redis2.sock .port: 6381
JSON file # dbid: 2 ,socket: /var/run/redis/redis1.sock .port: 6380
API get   # dbid: 2 ,socket: /var/run/redis/redis1.sock .port: 6380
JSON file # dbid: 5 ,socket: /var/run/redis/redis3.sock .port: 6382
API get   # dbid: 5 ,socket: /var/run/redis/redis3.sock .port: 6382
JSON file # dbid: 3 ,socket: /var/run/redis/redis2.sock .port: 6381
API get   # dbid: 3 ,socket: /var/run/redis/redis2.sock .port: 6381
JSON file # dbid: 5 ,socket: /var/run/redis/redis3.sock .port: 6382
API get   # dbid: 5 ,socket: /var/run/redis/redis3.sock .port: 6382
JSON file # dbid: 7 ,socket: /var/run/redis/redis4.sock .port: 6383
API get   # dbid: 7 ,socket: /var/run/redis/redis4.sock .port: 6383
JSON file # dbid: 6 ,socket: /var/run/redis/redis4.sock .port: 6383
API get   # dbid: 6 ,socket: /var/run/redis/redis4.sock .port: 6383
[       OK ] DBConnector.multi_db_test (0 ms)
[----------] 1 test from DBConnector (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (0 ms total)
[  PASSED  ] 1 test.

### CODE REVIEW

File: common/schema.h [hunk 1]
```diff
@@ -15,6 +15,8 @@ namespace swss {
 #define STATE_DB        6
 #define SNMP_OVERLAY_DB 7
 
+#define  DB_CONFIG_FILE "/var/run/redis/sonic-db/database_config.json"
```
Comment (qiluo-msft) [inline common/schema.h]: >DB_CONFIG_FILE [](start = 9, length = 14)

This file is about database schema, config file is not related.

File: common/table.h [hunk 2]
```diff
@@ -180,6 +180,7 @@ class Table : public TableBase, public TableEntryEnumerable {
      * 2) "Ethernet0,Ethernet4,...
      * */
     std::string stripSpecialSym(const std::string &key);
+    std::string m_sha;
```
Comment (qiluo-msft) [inline common/table.h]: >m_sha [](start = 16, length = 5)

m_shaDump

File: common/dbconnector.cpp [hunk 3]
```diff
@@ -4,14 +4,46 @@
...
 
 namespace swss {
 
+bool RedisConfig::parseRedisDBConfig(string file) {
+
+    ifstream i(file);
+    if (i.good()) {
+        json j;
+        i >> j;
+        for(auto it = j["INSTANCES"].begin(); it!= j["INSTANCES"].end(); it++) {
```
Comment (jleveque) [inline common/dbconnector.cpp]: Move opening curly brace to its own line to match the prevailing style.

Also add a space between `for` and the opening parenthesis.

File: common/dbconnector.cpp [hunk 4]
```diff
@@ -4,14 +4,46 @@
...
 
 #include "common/dbconnector.h"
 #include "common/redisreply.h"
 
+using json = nlohmann::json;
 using namespace std;
 
 namespace swss {
 
+bool RedisConfig::parseRedisDBConfig(string file) {
```
Comment (jleveque) [inline common/dbconnector.cpp]: Move opening curly brace to its own line to match the prevailing style.

File: common/dbconnector.cpp [hunk 5]
```diff
@@ -4,14 +4,46 @@
...
+        json j;
+        i >> j;
+        for(auto it = j["INSTANCES"].begin(); it!= j["INSTANCES"].end(); it++) {
+           string instName = it.key();
+           string sockPath = it.value().at("socket");
+           int port = it.value().at("port");
+           m_redis_info[instName] = {sockPath, port};
+        }
+
+        for(auto it = j["DATABASES"].begin(); it!= j["DATABASES"].end(); it++) {
```
Comment (jleveque) [inline common/dbconnector.cpp]: Move opening curly brace to its own line to match the prevailing style.

Also add a space between `for` and the opening parenthesis.

File: tests/redis_multi_db_ut.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    cout<<"Update : load local db config file, isInit = "<<RedisConfig::isInit()<<endl;
+    EXPECT_EQ(RedisConfig::isInit(), true);
+
+    // parse config file
+    ifstream i(file);
+    if (i.good()) {
+        json j;
+        i >> j;
+        unordered_map<string, pair<string, int>> m_redis_info;
+        for(auto it = j["INSTANCES"].begin(); it!= j["INSTANCES"].end(); it++) {
```
Comment (jleveque) [inline tests/redis_multi_db_ut.cpp]: Move opening curly brace to its own line to match the prevailing style.

Also add a space between `for` and the opening parenthesis.

File: tests/redis_multi_db_ut.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+           string instName = it.key();
+           string sockPath = it.value().at("socket");
+           int port = it.value().at("port");
+           m_redis_info[instName] = {sockPath, port};
+        }
+
+        for(auto it = j["DATABASES"].begin(); it!= j["DATABASES"].end(); it++) {
```
Comment (jleveque) [inline tests/redis_multi_db_ut.cpp]: Move opening curly brace to its own line to match the prevailing style.

Also add a space between `for` and the opening parenthesis.

File: common/table.cpp [hunk 8]
```diff
@@ -41,6 +41,8 @@ Table::Table(RedisPipeline *pipeline, const string &tableName, bool buffered)
     , m_pipeowned(false)
     , m_pipe(pipeline)
 {
+    std::string m_luaScript = loadLuaScript("table_dump.lua");
```
Comment (qiluo-msft) [inline common/table.cpp]: >m_luaScript [](start = 16, length = 11)

luaScript or luaDump, please don't use 'm_' for local variables.

File: common/consumerstatetable.cpp [hunk 9]
```diff
@@ -15,6 +15,10 @@ ConsumerStateTable::ConsumerStateTable(DBConnector *db, const std::string &table
     : ConsumerTableBase(db, tableName, popBatchSize, pri)
     , TableName_KeySet(tableName)
 {
+
+    std::string m_luaScript = loadLuaScript("consumer_state_table_pops.lua");
```
Comment (qiluo-msft) [inline common/consumerstatetable.cpp]: Remove extra empty line

Comment (qiluo-msft) [inline common/consumerstatetable.cpp]: >m_luaScript [](start = 16, length = 11)

luaScript or luaPop, please don't use 'm_' for local variables.

File: common/consumerstatetable.h [hunk 10]
```diff
@@ -14,6 +14,9 @@ class ConsumerStateTable : public ConsumerTableBase, public TableName_KeySet
 
     /* Get multiple pop elements */
     void pops(std::deque<KeyOpFieldsValuesTuple> &vkco, const std::string &prefix = EMPTY_PREFIX);
+
+private:
+    std::string m_sha;
```
Comment (qiluo-msft) [inline common/consumerstatetable.h]: m_shaPop

File: common/consumertable.cpp [hunk 11]
```diff
@@ -18,6 +18,9 @@ ConsumerTable::ConsumerTable(DBConnector *db, const string &tableName, int popBa
     : ConsumerTableBase(db, tableName, popBatchSize, pri)
     , TableName_KeyValueOpQueues(tableName)
 {
+    std::string m_luaScript = loadLuaScript("consumer_table_pops.lua");
```
Comment (qiluo-msft) [inline common/consumertable.cpp]: >m_luaScript [](start = 16, length = 11)

luaScript or luaPop, please don't use 'm_' for local variables.

File: common/consumertable.h [hunk 12]
```diff
@@ -17,6 +17,9 @@ class ConsumerTable : public ConsumerTableBase, public TableName_KeyValueOpQueue
 
     /* Get multiple pop elements */
     void pops(std::deque<KeyOpFieldsValuesTuple> &vkco, const std::string &prefix = EMPTY_PREFIX);
+
+private:
+    std::string m_sha;
```
Comment (qiluo-msft) [inline common/consumertable.h]: >m_sha [](start = 16, length = 5)

m_shaPop

File: tests/redis_multi_db_ut_config/database_config.json [hunk 13]
```diff
@@ -0,0 +1,63 @@
+{
+    "INSTANCES": {
+        "redis":{
+            "port": 6379,
```
Comment (qiluo-msft) [inline tests/redis_multi_db_ut_config/database_config.json]: >port [](start = 13, length = 4)

"hostname" should be part of the config file.

File: tests/redis_multi_db_ut_config/database_config.json [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            "socket": "/var/run/redis/redis.sock"
+        },
+        "redis1":{
+            "port": 6380,
+            "socket": "/var/run/redis/redis1.sock"
```
Comment (qiluo-msft) [inline tests/redis_multi_db_ut_config/database_config.json]: >socket [](start = 13, length = 6)

unixpath or unixsocketpath

File: common/dbconnector.h [hunk 15]
```diff
@@ -22,6 +44,8 @@ class DBConnector
      */
     DBConnector(int dbId, const std::string &hostname, int port, unsigned int timeout);
     DBConnector(int dbId, const std::string &unixPath, unsigned int timeout);
+    DBConnector(const std::string &hostname, int dbId, unsigned int timeout); //TCP PORT
```
Comment (qiluo-msft) [inline common/dbconnector.h]: >hostname [](start = 35, length = 8)

hostname should be part of the config file, then we don't need this overload function. RedisConfig may be responsible to select port or unix socket (default).

File: common/schema.h (hunk 1 above)
Comment (qiluo-msft) [inline common/schema.h]: >  [](start = 8, length = 1)

One more extra blank

File: common/dbconnector.h [hunk 16]
```diff
@@ -22,6 +44,8 @@ class DBConnector
...
+    DBConnector(int dbId, unsigned int timeout); //UNIX SOCKET
```
Comment (qiluo-msft) [inline common/dbconnector.h]: >dbId [](start = 20, length = 4)

dbId is not future proof.
Current your implementation assume the dbID(s) are unique among all the redis instances. In future the assumption may be removed, and we may reuse dbID for different databases in multiple redis instances. The function may be better design to index by a database name string instead of dbID.

File: common/dbconnector.cpp [hunk 17]
```diff
@@ -63,6 +95,40 @@ DBConnector::DBConnector(int dbId, const string& unixPath, unsigned int timeout)
     select(this);
 }
 
+DBConnector::DBConnector(const string& hostname, int dbId, unsigned int timeout) :
+    m_dbId(dbId)
+{
+    struct timeval tv = {0, (suseconds_t)timeout * 1000};
+
+    if (timeout)
+        m_conn = redisConnectWithTimeout(hostname.c_str(), RedisConfig::getDbport(dbId), tv);
```
Comment (qiluo-msft) [inline common/dbconnector.cpp]: >RedisConfig::getDbport(dbId) [](start = 59, length = 28)

DBConnector depends on the class RedisConfig's static members.

It is better to define a non-static RedisConfig object, and a function DBConnector RedisConfig::getDBConnector(const string& dataabse).

We already discussed this offline. DBConnector should not care about config file in concept.

**Comment by dzhangalibaba** [comment]
@qiluo-msft @jleveque  modified as requested, please take a look.

File: common/table.h (hunk 2 above)
Comment (qiluo-msft) [reply common/table.h]: It is for dumping, not for popping.

---
In reply to: [325908837](https://github.com/Azure/sonic-swss-common/pull/301#discussion_r325908837) [](ancestors = 325908837)

File: common/table.cpp (hunk 8 above)
Comment (qiluo-msft) [reply common/table.cpp]: dump?

---
In reply to: [325909420](https://github.com/Azure/sonic-swss-common/pull/301#discussion_r325909420) [](ancestors = 325909420)

File: common/table.h (hunk 2 above)
Comment (dzhangalibaba) [reply common/table.h]: yes it is Dump , corrected.

File: common/dbconnector.h [hunk 18]
```diff
@@ -3,11 +3,33 @@
...
+#include <utility>
 
 #include <hiredis/hiredis.h>
 
 namespace swss {
 
+class SonicDBConfig
+{
+public:
+    static void initialize(const std::string &file);
```
Comment (qiluo-msft) [inline common/dbconnector.h]: >file [](start = 46, length = 4)

Add a default value, so use can call initialize() without a parameter.

File: common/dbconnector.cpp [hunk 19]
```diff
@@ -4,14 +4,87 @@
...
+{
+    if (m_init)
+        throw runtime_error("SonicDBConfig already initialized");
+
+    ifstream i(file);
+    if (i.good())
+    {
+        json j;
+        i >> j;
+        for (auto it = j["INSTANCES"].begin(); it!= j["INSTANCES"].end(); it++)
```
Comment (qiluo-msft) [inline common/dbconnector.cpp]: >j["INSTANCES"] [](start = 23, length = 14)

If the file is corrupted, the index operator will throw. Better to catch it and throw a meaningful exception with nested exception.

Comment (dzhangalibaba) [reply common/dbconnector.cpp]: do we need to handle nested_exception? Or just throw_with_nested(runtime_error("Parse config file failed)")); is fine?

Comment (qiluo-msft) [reply common/dbconnector.cpp]: I don't understand your 2 questions.

To clarify my point, your current code will throw out_of_index exception to user, which is true but not easy to read. You can catch it and wrap it into another exception which indicates config file syntax error, and rethrow. All the stand exception has a constructor which you could feed an inner exception (out_of_index in this case) as a parameter. ref: https://en.cppreference.com/w/cpp/error/exception/exception

Comment (dzhangalibaba) [reply common/dbconnector.cpp]: I am still confused what you want. I currently understanding is not using  throw_with_nested().
do you mean:
try {
    array[out_of_index]
} 
catch (exception &e)
{
    throw runntime_error("config file syntax error : " + string(e.what()))
}

Comment (dzhangalibaba) [reply common/dbconnector.cpp]: if this is not what you mean, we can have a quick call to make it clear in the morning.

File: common/dbconnector.cpp [hunk 20]
```diff
@@ -4,14 +4,94 @@
...
+            for (auto it = j["DATABASES"].begin(); it!= j["DATABASES"].end(); it++)
+            {
+               string dbName = it.key();
+               string instName = it.value().at("instance");
+               int dbId = it.value().at("id");
+               m_db_info[dbName] = {instName, dbId};
+            }
+            m_init = true;
+        }
+        catch (exception &e)
```
Comment (qiluo-msft) [inline common/dbconnector.cpp]: >exception [](start = 15, length = 9)

Be as specific as you can. I guess it is `out_of_range`. Please have a test.

Comment (dzhangalibaba) [reply common/dbconnector.cpp]: The exceptions is not out_of_range.  Since  "j"  is a json object, j["INDEX"] return NULL if there is no "INDEX" field in json object. The exception throw nwhen we try to do  j[INDEX].begin(), because NULL has no begin() to call. the exception type is "json::invalid_iterator". will update codes soon

Comment (dzhangalibaba) [reply common/dbconnector.cpp]: Just FYI,

we are still using json.hpp 2.0 version on Sonic which has no json::invalid_iterator added .
so it is throwing  domain_error for that null iterator case, I tested on Sonic,.

On my local machine, I am using json.hpp version 3.0 , that is json::invalid_iterator exception.

File: common/table.cpp [hunk 21]
```diff
@@ -41,6 +41,8 @@ Table::Table(RedisPipeline *pipeline, const string &tableName, bool buffered)
     , m_pipeowned(false)
     , m_pipe(pipeline)
 {
+    std::string luaScript = loadLuaScript("table_dump.lua");
+    m_shaDump = pipeline->loadRedisScript(luaScript);
```
Comment (qiluo-msft) [inline common/table.cpp:45]: >loadRedisScript [](start = 26, length = 15)

This is not efficient. Could you load on demand?

Comment (dzhangalibaba) [reply common/table.cpp:45]: > > loadRedisScript [](start = 26, length = 15)
> 
> This is not efficient. Could you load on demand?

PR raised , also handle another similar case introduced by kcudnik
